runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -243,19 +252,53 @@ function TSHighlighter:get_query(lang)
   return self._queries[lang]
 end
 
+--- @param match table<integer,TSNode[]>
+--- @param bufnr integer
+--- @param capture integer
+--- @param metadata vim.treesitter.query.TSMetadata
+--- @return string?
+local function get_url(match, bufnr, capture, metadata)
+  ---@type string|number|nil
+  local url = metadata[capture] and metadata[capture].url
+
+  if type(url) ~= 'number' then
+    return
+  end
+
+  if not match or not match[url] then
+    return
+  end
+
+  -- Assume there is only one matching node. If there is more than one, take the URL
+  -- from the first.
+  local other_node = match[url][1]
+
+  return vim.treesitter.get_node_text(other_node, bufnr, {
+    metadata = metadata[url],
+  })
+end
+
+--- @param capture_name string
+--- @return boolean?, integer
+local function get_spell(capture_name)
+  if capture_name == 'spell' then
+    return true, 0
+  elseif capture_name == 'nospell' then
+    -- Give nospell a higher priority so it always overrides spell captures.
+    return false, 1
+  end
+  return nil, 0
+end
+
 ---@param self vim.treesitter.highlighter
 ---@param buf integer
 ---@param line integer
 ---@param is_spell_nav boolean
 local function on_line_impl(self, buf, line, is_spell_nav)
-  -- Track the maximum pattern index encountered in each tree. For subsequent
-  -- trees, the subpriority passed to nvim_buf_set_extmark is offset by the
-  -- largest pattern index from the prior tree. This ensures that extmarks
-  -- from subsequent trees always appear "on top of" extmarks from previous
-  -- trees (e.g. injections should always appear over base highlights).
-  local pattern_offset = 0
-
   self:for_each_highlight_state(function(state)
+    -- Use the injection level to offset the subpriority passed to nvim_buf_set_extmark
+    -- so injections always appear over base highlights.
+    local pattern_offset = state.level * 1000
```

Comment:
I avoided using some constant scaling factor intentionally because you run the risk of the offset exceeding this factor. It's true that it's _unlikely_ that there will ever be more than 1000 patterns in a query, I still think fixed scale factors like this are not robust (it's the "640K RAM should be enough for anyone" mistake all over again). That is the reason I used the maximum index from the previous tree.

Reply:
It's not ideal, but I don't know what else we can do.

Reply:
Tbh this really isn't the same as "640K RAM should be enough for anyone" . We can increase this number without consequence as we see fit. If we run out of room, we can arbitrarily make more room.

Reply:
Possible alternative: Make `TSQuery:inspect` return the number of patterns in the query and pre-compute the offset in `on_win`.